### PR TITLE
Move the simulator into the main code

### LINKED ISSUE #35: Move the simulator into the main code
Currently, the simulator lives in `examples/`. This means that:
- it is not run by `cargo test`;
- it cannot be used by tests;
- it cannot be used by FoxBox.

We should move the code of the simulator into its module of `src/` and use it to write tests.

Once this transformation is complete, `examples/simulator.rs` should pretty much be reduced to its current `fn main()`.

### PR DESCRIPTION
**Pull request by JohanLorenzo**
Fixes #35

### CODE REVIEW

File: src/simulator/api.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+use std::sync::Arc;
+
+use serde::ser::{Serialize, Serializer};
+use serde::de::{Deserialize, Deserializer};
+
+use simulator::instruction::Op;
+
+
+#[derive(Debug)]
+pub enum Update {
```
Comment (JohanLorenzo) [inline src/simulator/api.rs]: Not sure if the next 2 enums should be in a seperate file.

**Comment by JohanLorenzo** [comment]
It took a bit more time than I expected yesterday/ What basically changed in this PR:
* Imports
* Visibility of functions and structs
* Dead code allowed if regular build is done

r? @Yoric

File: src/simulator/instruction.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+use foxbox_taxonomy::selector::*;
+use foxbox_taxonomy::api::{WatchEvent, WatchOptions, Error};
+
+#[derive(Serialize, Deserialize, Debug)]
+/// Instructions given to the simulator.
+pub enum Instruction {
+    AddNodes(Vec<Node>),
+    AddGetters(Vec<Channel<Getter>>),
+    AddSetters(Vec<Channel<Setter>>),
+    InjectGetterValue{id: Id<Getter>, value: Value},
```
Comment (Yoric) [inline src/simulator/instruction.rs]: Apparently, we're supposed to add spaces, as follows:
`InjectGetterValue { id: ... Value },`

File: src/simulator/instruction.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+/// Operations internal to the simulator.
+pub enum Op {
+    AddNodes(Vec<Node>),
+    AddGetters(Vec<Channel<Getter>>),
+    AddSetters(Vec<Channel<Setter>>),
+    AddWatch{options: Vec<WatchOptions>, cb: Box<Fn(WatchEvent) + Send + 'static>},
```
Comment (Yoric) [inline src/simulator/instruction.rs]: Spaces here, too.

File: src/simulator/instruction.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+/// Instructions given to the simulator.
+pub enum Instruction {
+    AddNodes(Vec<Node>),
+    AddGetters(Vec<Channel<Getter>>),
+    AddSetters(Vec<Channel<Setter>>),
+    InjectGetterValue{id: Id<Getter>, value: Value},
+}
+
+/// Operations internal to the simulator.
+pub enum Op {
```
Comment (Yoric) [inline src/simulator/instruction.rs]: There is an inconsistency in this module.

If this module is public, then both `Instruction` and `Op` will end up public. We don't want `Op` to be public.
If this module is private, then both `Instruction` and `Op` will be private. We don't want `Instruction` to be private.

Consequently, I believe that we should not put `Instruction` and `Op` in the same module. `Instruction` should end in a `pub mod`, while `Op` should not.

File: src/simulator/api.rs (hunk 1 above)
Comment (Yoric) [inline src/simulator/api.rs]: Nah, leave `Update` here.

File: src/simulator/api.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                result: internal_result
+            });
+            (setter.id.clone(), result)
+        }).collect();
+        cb(results)
+    }
+}
+
+#[derive(Clone)]
+pub struct APIFrontEnd {
```
Comment (Yoric) [inline src/simulator/api.rs]: Could you document this?

Also, I'm not sure that `APIFrontEnd` is a very good name. Maybe `DevicesSimulation`?

File: src/simulator/api.rs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    }
+}
+impl Default for APIFrontEnd {
+    fn default() -> Self {
+        panic!("WTF are we doing calling default() for the front-end?");
+    }
+}
+
+impl APIFrontEnd {
+    pub fn new<F>(cb: F) -> Self
```
Comment (Yoric) [inline src/simulator/api.rs]: Also, this function should be documented.

File: src/simulator/instruction.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    AddNodes(Vec<Node>),
+    AddGetters(Vec<Channel<Getter>>),
+    AddSetters(Vec<Channel<Setter>>),
+    AddWatch{options: Vec<WatchOptions>, cb: Box<Fn(WatchEvent) + Send + 'static>},
+    SendValue{selectors: Vec<SetterSelector>, value: Value, cb: Box<Fn(Vec<(Id<Setter>, Result<(), Error>)>) + Send>},
+    InjectGetterValue{id: Id<Getter>, value: Value},
+}
+
+impl Instruction {
+    pub fn as_op(self) -> Op {
```
Comment (Yoric) [inline src/simulator/instruction.rs]: This probably doesn't need to be `pub`.

File: src/simulator/instruction.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+use foxbox_taxonomy::util::Id;
+use foxbox_taxonomy::values::*;
+use foxbox_taxonomy::devices::*;
+use foxbox_taxonomy::selector::*;
+use foxbox_taxonomy::api::{WatchEvent, WatchOptions, Error};
+
+#[derive(Serialize, Deserialize, Debug)]
+/// Instructions given to the simulator.
+pub enum Instruction {
```
Comment (Yoric) [inline src/simulator/instruction.rs]: This needs a few more words, to mention that, in practice, this is how clients are going to script the simulator.

**Comment by Yoric** [comment]
Thanks for the patch. I have a few improvements in mind, while you're at it :)

**Comment by Yoric** [comment]
As Github doesn't offer a good way to track updates, I'll close the PR. Please reopen it once you want me to look at it again.

**Pull request closed without merging**